cranelift/assembler-x64/meta/src/dsl/custom.rs

```diff
@@ -0,0 +1,85 @@
+use core::fmt;
+use std::ops::BitOr;
+
+#[derive(PartialEq, Debug)]
+pub enum CustomOperation {
+    Visit,
+    Display,
+    None,
+}
+
+impl fmt::Display for CustomOperation {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Debug::fmt(self, f)
+    }
+}
+
+impl BitOr for CustomOperation {
+    type Output = Custom;
+    fn bitor(self, rhs: Self) -> Self::Output {
+        assert_ne!(self, rhs, "duplicate custom operation: {self:?}");
+        Custom(vec![self, rhs])
+    }
+}
+
+impl BitOr<CustomOperation> for Custom {
+    type Output = Custom;
+    fn bitor(mut self, rhs: CustomOperation) -> Self::Output {
+        assert!(
+            !self.0.contains(&rhs),
+            "duplicate custom operation: {rhs:?}"
+        );
+        self.0.push(rhs);
+        self
+    }
+}
+
+#[derive(PartialEq)]
+pub struct Custom(pub Vec<CustomOperation>);
```

Comment:
```suggestion
pub struct Custom(Vec<CustomOperation>);
```

No need for this to be `pub` if you have `iter` below.